README.md
# PWA Install Prompt

![screenshot](screenshot.png)

Prompt users to add your PWA to their home screen, since Apple won’t.

By default, this module checks if the user has not visited the site before, or if the last time they visited the site was more than 90 days ago, and if they are using iOS. If all conditions are met, then the prompt is shown. The conditions that get checked can be overridden in order to use this prompt for custom scenarios.

*[Inspired by Scott Batson @ DockYard](https://dockyard.com/blog/2017/09/27/encouraging-pwa-installation-on-ios)*

## Demo

https://jacobdb.github.io/pwa-install-prompt/demo/

## Installation

### via npm

```sh
npm install pwa-install-prompt --save
```

### via yarn

```sh
yarn add pwa-install-prompt
```

## Usage

```js
import pwaInstallPrompt from "pwa-install-prompt";

new pwaInstallPrompt(".pwa-install-prompt__container", {
    active_class: "is-active",
    closer: ".pwa-install-prompt__overlay",
    condition: null,
    expires: 180,
    show_after: 90,
    on: {
        beforeOpen: function () {
            console.log("before open!");
        },
        afterOpen: function () {
            console.log("after open!");
        },
        beforeClose: function () {
            console.log("before close!");
        },
        afterClose: function () {
            console.log("after close!");
        },
    }
});
```

## Default Markup

```html
<div class="pwa-install-prompt__container">
    <button class="pwa-install-prompt__overlay">Close</button>
    <div class="pwa-install-prompt">
        <div class="pwa-install-prompt__icon__container">
            <img class="pwa-install-prompt__icon" src="/path/to/icon/file.png" alt="{app_name}" />
        </div>
        <div class="pwa-install-prompt__content">
            <h3 class="pwa-install-prompt__title">Install {app_name}</h3>
            <p class="pwa-install-prompt__text">Install this application on your home screen for quick and easy access when you’re on the go.</p>
            <p class="pwa-install-prompt__guide">Just tap <svg class="pwa-install-prompt__guide__icon" viewBox="0 0 128 128" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink"><title>Share</title><path fill="#007AFF" d="M48.883,22.992L61.146,10.677L61.146,78.282C61.146,80.005 62.285,81.149 64,81.149C65.715,81.149 66.854,80.005 66.854,78.282L66.854,10.677L79.117,22.992C79.693,23.57 80.256,23.853 81.114,23.853C81.971,23.853 82.534,23.57 83.11,22.992C84.25,21.848 84.25,20.125 83.11,18.981L65.997,1.794C65.715,1.511 65.421,1.215 65.139,1.215C64.563,0.932 63.718,0.932 62.861,1.215C62.579,1.498 62.285,1.498 62.003,1.794L44.89,18.981C43.75,20.125 43.75,21.848 44.89,22.992C46.029,24.149 47.744,24.149 48.883,22.992ZM103.936,35.32L81.114,35.32L81.114,41.053L103.936,41.053L103.936,121.27L24.064,121.27L24.064,41.053L46.886,41.053L46.886,35.32L24.064,35.32C20.928,35.32 18.355,37.904 18.355,41.053L18.355,121.27C18.355,124.419 20.928,127.003 24.064,127.003L103.936,127.003C107.072,127.003 109.645,124.419 109.645,121.27L109.645,41.053C109.645,37.891 107.072,35.32 103.936,35.32Z" /></svg> then “Add to Home Screen”</p>
        </div>
    </div>
</div>
```

## Options

| Option         | Type       | Default                                 | Description                                                 |
|----------------|------------|-----------------------------------------|-------------------------------------------------------------|
| `active_class` | `string`   | `is-active`                             | Class name to be added when the prompt is visible           |
| `closer`       | `string`   | `.pwa-install-prompt__overlay`          | Query selector to target to handle closing the prompt       |
| `condition`    | `function` | `null`                                  | Set a custom condition to determine when to show the prompt |
| `expires`      | `number`   | `180`                                   | Number of days before the `last_visit` cookie expires       |
| `show_after`   | `number`   | `90`                                    | Number of days to wait before showing the prompt again      |
| `on`           | `object`   | `null`                                  | Register event handlers                                     |

## Events

Events can be assigned in two ways:

1. Using on parameter on PWA Install Prompt initialization:

        var prompt = new pwaInstallPrompt(".pwa-install-prompt__container", {
            // ...
            on: {
                open: function () {
                    console.log("prompt opened!");
                },
            },
        });

2. Using on method after PWA Install Prompt initialization.

        var prompt = new pwaInstallPrompt(".pwa-install-prompt__container", {
            // ...
        });

        prompt.on("open", function () {
            console.log("prompt opened!");
        });

| Name          | Description                                       |
|---------------|---------------------------------------------------|
| `beforeOpen`  | Event will be fired right before prompt is opened |
| `afterOpen`   | Event will be fired right after prompt is opened  |
| `beforeClose` | Event will be fired right before prompt is closed |
| `afterClose`  | Event will be fired right after prompt is closed  |

## Contributors

- [Jacob Bearce](https://github.com/JacobDB) &ndash; [Commits](https://github.com/JacobDB/pwa-install-prompt/commits?author=JacobDB)
- [Sergio Arbeo](https://github.com/Serabe) &ndash; [Commits](https://github.com/JacobDB/pwa-install-prompt/commits?author=Serabe)


CHANGELOG.md
# Changelog

## v1.1.1

- Correct day count; aece95b8ac369583349501ade8038edd2ba81302
- Correct day counter information; 3005a0f9d9763b3fd28aaa40421016f27a67ef4c
- Fix cookie SameSite policy; 2164459a89ba668d5c93c0c15961c3025c3dbf2f

## v1.1.0

- Change cookie name to `pwa_install_prompt_last_visit` to prevent conflicts; Close #8; fe32f7c494fc7e16ea7370844d29b7c04a136428
- Extend default `show_after` to 90 days, matching Chrome's native install prompt; Close #11; db5ff6d541d156ef37cc0582927efa446574aa25
- Add beforeOpen and beforeClose events; Close #10; 23f10cac06ed4ea09c3c9f9c63687005304ad980
- Fix usage example; 1597e80e2b1177c4ccd69c10990bbb5a6d6d625f

## v1.0.2

- Fix check for iOS so that iPads are included; 11430c36ef7a49e83a001b4104f888ec04fb1be7

## v1.0.1

- Remove testing events from demo; 839fd85b190ed334b01eae702107bbf4000c86b8
- Correct package name in events documentation; be775e2a5cfe430b34d91891c2eab4b38f255174

## v1.0.0

- Add custom events; Close #2; 7ed02e9322b19e34127479e83efa2fee21abd695
- Add icon container to default markup; c71e41c8a6b53e092825230b92b79c7a0b8dd3ca

## v0.1.14

- Fix example markup in the README; 1256d43822f37a24f399f8c93ebe54de47977cbf
- Add line break to example script in README; 9f5e1e119d6fc8c1f7471a2317083b43bd707fcd
- Organize options a bit better; 73c80a085426c897766f632f6f8520277c767bc4
- Adjust the demo so it's more like a production environment; 69880fa8dd064d24bfc915a7694194b4369d3545
- Add links to GitHub and NPM to the demo; b515b399d29d7fbfa06ad535fe2641b26623239c
- Curly apostrophes; eab35155b3eeed46b03ed5f349105c94c4f3201c
- Add 'open' and 'close' methods; adc3bd7acea0aa0a21131eebab2971d62072c1f4
- Add "ghost" icons next to the main app icon; 9052ef0c117679c2a9144d5030f3e2b44b48349d

## v0.1.13

- Change to a single closing element; 5bd29529347e39e291dfac4a24a9db11854ad1e4

## v0.1.12

- Make the demo behave more like it would in production; a867fe68da14d8ee7a6889743ae00626f4444e76
- Delay initialization for the demo; a96bfabdd7c816e56587a5932d0dbeecf7f3a558
- Add a note clarifying how the demo works; 3172f0ee1cf980fb5e26d358cbc245916c6c467c
- Fix link to DockYard post; PR #5; 2e7d720c632a2e9c9cc30cab1785476e8f43be7f
- Add contributors list; 091b8d7957717e044c53e3ea5d0ddc314a705de3

## v0.1.11

- Add transition effects; b85a09dabfe2f34c76f5c2369b367afc75ffdeb7

## v0.1.10

- Make the description a bit more clear; c1abc1f08dd7c9f0563e9f7fd807cadc527e91cf
- Document `expires` option; 1259e6556b1d9768b9b12d5368f33822478c5d19

## v0.1.9

- Fix iOS share icon blue to match Apple's HID spec; 7c15c1c4c4a6ae9ae8a061df8d3d3f8797f12079
- Add screenshot to README; d8e71cd929f2e9745f732008f137c64f741742bc

## v0.1.8

- Change from arrow functions to standard to fix 'not a constructor' error; 0a0e0ac1730e22c8dbabd21e09e214a7b974bef1

## v0.1.7

- Use for (...) instead of forEach; e497b0d9a143683cc74584b8ebb37a9f5feb01f1

## v0.1.6

- Change main function back from an arrow function; c34a03b6ed0ebb628e612ab817f24b91a1298423

## v0.1.5

- Change over to all arrow functions; 37017c7dd62a2c596a472a00920a3861c32f587c
- Use Array.indexOf instead of Array.includes; 26e3a28992afd4889a6b03156e3ba0221b4291e6

## v0.1.4

- Fix a few variable names; e8f9d6e81b602c4ffd62a026d67033e08e192b2b

## v0.1.3

- Add FontAwesome icon credit; 226c25768ca2435dfc724f48009d58a8a98cd6f0
- Add stylesheet to package.json; 3d87fee081ae7c0ea272c8a675c54e05d9d54d9d
- Standardize SVG sizes; 466deeca92561ce609be658e35d3316cb31875c8
- Add alt text for share icon; Close #4; 17f0fedd0aff7c214f284d1ca278344ea917e087
- Allow the user to set a custom condition; 68ccdf33e1b10df1f5cf979bccb3d57eaee6fd82
- Rename from ios-install-prompt to pwa-install-prompt; 7a8029b66c59b5725fb657e435a0132cab46f4e5

## v0.1.2

- Make it more clear how options are to be used; c1065456ac7a11a77b6c6b44c4617b75946345a5
- Use default values for example options; b8f0f88423421ad4f30cba7fd7122b1c41309b4a
- Fix the build number; 4e558d94ab58454b496c0a2c98cce8693a7ad137

## v0.1.1

- Add inspiration and demo links; 602aef245e623babde93c8c8c0ba6bd41163c582
- Add note to remove is-active class for production; 04c213a9cf172c5d897bb1c0af299f3c4e1445b9
- Allow the overlay to be closed even if the prompt wasn't shown by the script; e844f401c347b27e3c7959f52b22f2b6519d1302

## v0.1.0

- Initial Commit; 645fe3af962a393f09b9a4bae4ffa4a569c4e4cb


script.js
/**
 * PWA Install Prompt 1.0.2
 * Prompt users to add your PWA to their home screen, since Apple won’t.
 * https://github.com/JacobDB/pwa-install-prompt/
 *
 * Copyright 2018 Jacob Bearce
 *
 * Released under the MIT License
 *
 * Released on: November 6, 2018
 */

(function (global, factory) {
    typeof exports === "object" && typeof module !== "undefined" ? module.exports = factory() :
    typeof define === "function" && define.amd ? define(factory) :
    (global.pwaInstallPrompt = factory());
}(this, (function () {
    const DEFAULTS = {
        active_class: "is-active",
        closer:       ".pwa-install-prompt__overlay",
        condition:    null,
        container:    ".pwa-install-prompt__container",
        expires:      180,
        show_after:   90,
    };

    const NEEDS_TO_SEE_PROMPT = function (options) {
        // allow user to set a custom condition that overrides below
        if (options.condition !== null) {
            return options.condition;
        }

        // never show it if the app was launched standalone
        if (navigator.standalone) {
            return false;
        }

        // get the current date and the date the user last visited
        const NOW  = Date.now();
        const THEN = document.cookie.match(/pwa_install_prompt_last_visit=([0-9]+)/);

        // update the pwa_install_prompt_last_visit cookie with Date.now()
        document.cookie = "pwa_install_prompt_last_visit=" + NOW + "; max-age=" + (60 * 60 * 24 * options.expires) + "; SameSite=None; Secure";

        // get the number of days since the user last visited and check if they're using iOS
        const DAYS = THEN && THEN.length >= 2 ? Math.ceil((parseInt(THEN[1]) - NOW) / (1000 * 60 * 60 * 24)) : NaN;
        const IOS  = ["iPad", "iPhone", "iPod"].indexOf(navigator.platform) > -1;

        // show if the cookie isn't set, or last visit is > show_after, and it's iOS
        return (isNaN(DAYS) || DAYS > options.show_after) && IOS;
    };

    const INSTALL_PROMPT = (function (container = DEFAULTS.container, options = {}) {
        const PROMPT = this;

        options.active_class = options.active_class || DEFAULTS.active_class;
        options.closer       = options.closer       || DEFAULTS.closer;
        options.condition    = options.condition    || DEFAULTS.condition;
        options.container    = container;
        options.expires      = options.expires      || DEFAULTS.expires;
        options.show_after   = options.show_after   || DEFAULTS.show_after;
        options.on           = options.on           || false;

        // store the arguments for later
        PROMPT.container       = container;
        PROMPT.options         = options;
        PROMPT.eventListeners = {};

        if (PROMPT.options.on) {
            // listen for each event specified
            Object.keys(PROMPT.options.on).forEach(function (event_name) {
                PROMPT.on(event_name, PROMPT.options.on[event_name]);
            });
        }

        // open if the conditional returns true
        if (NEEDS_TO_SEE_PROMPT(PROMPT.options)) {
            PROMPT.open();
        }

        const CLOSER = document.querySelector(PROMPT.options.closer);

        // close if the closer is clicked
        if (CLOSER) {
            CLOSER.addEventListener("click", function () {
                PROMPT.close();
            });
        }
    });

    INSTALL_PROMPT.prototype.on = function (events, handler) {
        const PROMPT = this;

        // prevent errors if a function wasn't passed
        if (typeof handler !== "function") { return PROMPT; }

        // set up eventListeners array to associate events with their specified handlers
        events.split(" ").forEach(function (event) {
            if (!PROMPT.eventListeners[event]) { PROMPT.eventListeners[event] = []; }
            PROMPT.eventListeners[event].push(handler);
        });

        return PROMPT;
    };

    INSTALL_PROMPT.prototype.emit = function () {
        const PROMPT = this;

        // prevent errors if no event listeners are specified
        if (!PROMPT.eventListeners) { return PROMPT; }

        // convert the events in to an array if they're not already
        const EVENTS = Array.isArray(arguments[0]) ? arguments[0] : arguments[0].split(" ");

        EVENTS.forEach(function (event) {
            // verify that the emitted event is being listend for
            if (PROMPT.eventListeners && PROMPT.eventListeners[event]) {
                // fire each handler for the event
                PROMPT.eventListeners[event].forEach(function (eventHandler) {
                    eventHandler.apply(PROMPT);
                });
            }
        });

        return PROMPT;
    };

    INSTALL_PROMPT.prototype.open = function () {
        const PROMPT    = this;
        const CONTAINER = document.querySelector(PROMPT.options.container);

        PROMPT.emit("beforeOpen");

        if (CONTAINER) {
            CONTAINER.classList.add(PROMPT.options.active_class);
        }

        PROMPT.emit("afterOpen");
    };

    INSTALL_PROMPT.prototype.close = function () {
        const PROMPT    = this;
        const CONTAINER = document.querySelector(PROMPT.options.container);

        PROMPT.emit("beforeClose");

        if (CONTAINER) {
            CONTAINER.classList.remove(PROMPT.options.active_class);
        }

        PROMPT.emit("afterClose");
    };

    return INSTALL_PROMPT;
})));
